### PR TITLE
feature: GH 67 south african reserve scraper

### PR DESCRIPTION
**Pull request by SFJohnson24**
added south african reserve scraper

### CODE REVIEW

File: scrapers/south_african_reserve.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+                file.write(response.content)
+            md5hash = get_md5hash(filepath)
+        
+        if parse_arguments().debug:
+            print(f"MD5hash for downloaded file: {md5hash}")
+
+        browser.close()
+    
+if __name__ == "__main__":
+    south_african_reserve_scraper()
```
Comment (nattiechan) [inline scrapers/south_african_reserve.py:44]: Need a new line at the end of the file

File: scrapers/utils.py [hunk 2]
```diff
@@ -111,7 +112,7 @@ def validate_and_get_download_link(base_url, download_link, check_content_pdf=Tr
         response = requests.head(absolute_url)
         content_type = response.headers.get('Content-Type', '')
         is_pdf = 'application/pdf' in content_type
-        if not is_pdf:
+        if not is_pdf and download_link[-4:] != '.pdf':
```
Comment (nattiechan) [inline scrapers/utils.py:115]: I requested a similar change in a different PR: https://github.com/sgoldfarb2/scraper-infra/pull/10#discussion_r1423104472

For the purpose of keeping changes clean and not co-mingling with each other, let's take this change out.

File: scrapers/utils.py [hunk 3]
```diff
@@ -2,14 +2,15 @@
...
 import inspect
-import requests
 import shutil
 import sys
 import tempfile
 from contextlib import contextmanager
 from datetime import datetime
 from pathlib import Path
 
+import requests
```
Comment (nattiechan) [inline scrapers/utils.py:12]: same here

**Review by nattiechan — CHANGES_REQUESTED** [review]
Before you address any comments regarding the code, I suspect the URL is incorrect.

With the URL you provided in the scraper, they only show circulars up to 2015, which does not seem right.

When I click "Bank Circulars" at the bottom right under `Publications Categories`, it opens a different page (https://www.resbank.co.za/en/home/publications/prudential-authority/banks-circulars) that seems to have more recent circulars available.

@sgoldfarb2 can we confirm which URL is correct?

File: .vscode/settings.json [hunk 4]
```diff
@@ -4,7 +4,7 @@
     "editor.defaultFormatter": "ms-python.black-formatter",
     "editor.formatOnSave": true,
     "editor.codeActionsOnSave": {
-      "source.organizeImports": true
```
Comment (nattiechan) [inline .vscode/settings.json:7]: I also noticed this change with newer versions of VS code. Can we double check what this change is about before committing to this change?

File: scrapers/south_african_reserve.py [hunk 5]
```diff
@@ -18,27 +21,39 @@ def south_african_reserve_scraper():
         page.goto(CIRCULAR_URL)
         session = requests.Session()
 
-        page.locator("button.cmp-accordion__button").all()[0].click()
-        pdf_link = page.locator("div.responsivegrid > div > div > div > p > b > a:visible").all()[-1]
-        link=pdf_link.get_attribute("href")
-        
+        # page.locator("button.cmp-accordion__button").all()[0].click()
+        # pdf_link = page.locator("div.responsivegrid > div > div > div > p > b > a:visible").all()[-1]
+        # link=pdf_link.get_attribute("href")
```
Comment (nattiechan) [inline scrapers/south_african_reserve.py:26]: Can we clean up unused code please?

File: scrapers/utils.py [hunk 6]
```diff
@@ -130,7 +130,7 @@ def validate_and_get_download_link(base_url, download_link, check_content_pdf=Tr
         response = requests.head(absolute_url)
         content_type = response.headers.get("Content-Type", "")
         is_pdf = "application/pdf" in content_type
-        if not is_pdf:
-            raise TypeError("Content-Type is not application/pdf.")
+        if not is_pdf and download_link.endswith(".pdf"):
+            raise TypeError("File extension does not match content type.")
```
Comment (nattiechan) [inline scrapers/utils.py:134]: Has this change been included in another PR?

**Review by nattiechan — CHANGES_REQUESTED** [review]

**Comment by nattiechan** [comment]
@SFJohnson24 have you confirmed with @sgoldfarb2 regarding the URL?

File: scrapers/south_african_reserve.py (hunk 5 above)
Comment (SFJohnson24) [reply scrapers/south_african_reserve.py]: removed

File: scrapers/utils.py (hunk 6 above)
Comment (SFJohnson24) [reply scrapers/utils.py:134]: this change is in the bank of ghana PR as well I believe

**Review by nattiechan — APPROVED** [review]
:tada:

**Review by jiedong111 — APPROVED** [review]
LGTM!

**Pull request merged**